tests/test_interpolation.py

```diff
@@ -753,3 +758,251 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
     OmegaConf.register_new_resolver("test", lambda x: x)
     cfg = OmegaConf.create({"x": "${test:'${missing}'}", "missing": None})
     assert cfg.x == "None"
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key", "expected_value", "expected_node_type"),
+    [
+        pytest.param(
+            User(name="Bond", age=SI("${cast:int,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="expected_type",
+        ),
+        pytest.param(
+            # This example specifically test the case where intermediate resolver results
+            # are not of the same type as the key.
+            User(name="Bond", age=SI("${cast:int,${drop_last:${drop_last:7xx}}}")),
```

Comment:
Side note, this would have looked somewhat nicer if we had () syntax support. Maybe for 2.2 :).

```
${cast:int, ${identity:'7'}
vs
${cast(int, ${identity('7'))}
```
I would also say that for a single argument, `:` is better.
So:
```
${cast(int, ${identity:'7')}
```

Reply:
> Can this be tested with identity instead? drop_last seems overly complex for what this test is doing.
> 
> ```python
>             User(name="Bond", age=SI("${cast:int,${identity:'7'}")),
> ```
> 
> Maybe even nest two identity calls (not sure what nesting twice is adding to the test).

I wanted to test explicitly the situation where there is an intermediate resolver output (by intermediate, I mean not the final result of the interpolation) that would be invalid if we tried to cast it to the node's type. The motivation was that my first implementation was doing the cast immediately after a resolver was called (even an intermediate one), which was incorrect (so I wanted a test to catch that).

If I just use `identity` then the string remains `"7"` which is actually valid when cast to the type of the node (integer). In that regard my comment motivating the test is not clear enough => I changed it in a99f776

I could also just remove that test if you feel like the extra complexity isn't worth it.

Reply:
I think a combination of casts can do it then :)

```python
User(name="Bond", age=SI("${cast:int,${cast:str,${identity:7}}}")),
```

Reply:
> Side note, this would have looked somewhat nicer if we had () syntax support. Maybe for 2.2 :).
> 
> ```
> ${cast:int, ${identity:'7'}
> vs
> ${cast(int, ${identity('7'))}
> ```
> 
> I would also say that for a single argument, `:` is better.
> So:
> 
> ```
> ${cast(int, ${identity:'7')}
> ```

I guess the version I like most is this:

```
${cast(int, identity('7'))}
```

As in, once we are in an interpolation we no longer need the ${}, we are already in function mode.

Reply:
> I think a combination of casts can do it then :)
> 
> ```python
> User(name="Bond", age=SI("${cast:int,${cast:str,${identity:7}}}")),
> ```

If I'm not mistaken the resolver outputs here are all either `7` (the integer) or `"7"` (the string) => both can be cast to int so that wouln't address the specific issue I wanted to pinpoint.

Reply:
Yeah, I guess I don't fully understand the error you are defending against.

Reply:
This was an error with the approach from https://github.com/omry/omegaconf/pull/506/files

In this approach, the call to `validate_and_convert()` happened immediately after a resolver was called, so in a chain of resolvers *all* outputs were validated (and possibly converted) according to the node's type. But that's wrong: it's only the final result that needs to be checked.